### PR TITLE
Check harder before enabling Accessibility support on Win32

### LINKED ISSUE #7208: Electron enables accessibility mode all the time with touch monitors (100% CPU usage in "Other")
* Electron version: 1.3.6 (win32-ia32 zip)
* Operating system: Windows 10.0.14393 Build 14393

Create a `main.js` file that loads a wikipedia page:

```js
var electron = require('electron')
var app = electron.app
var BrowserWindow = electron.BrowserWindow

var mainWindow;

function createWindow () {
	mainWindow = new BrowserWindow({width: 800, height: 600})
	mainWindow.loadURL(`https://en.wikipedia.org/wiki/Timi%C8%99oara`)
	mainWindow.webContents.openDevTools()
	mainWindow.on('closed', function () {
		mainWindow = null
	})
}

app.on('ready', createWindow)

app.on('window-all-closed', function () {
	if (process.platform !== 'darwin') {
		app.quit()
	}
})

app.on('activate', function () {
	if (mainWindow === null) {
		createWindow()
	}
})

```
* run `electron.exe main.js`
* scroll up and down rapidly by dragging the scrollbar
* watch CPU usage or Record with the Timeline
* **renderer process goes to 100% CPU usage**
* main process stays at 1-2% CPU usage
* both Chromium on a version before and Chrome on a version after do not have this problem.

## Electron v1.3.6 (process.versions.chrome = 52.0.2743.82)
image

## Chromium version 52.0.2743.0
image

## Chrome version 53.0.2785.113 m (64-bit)
image

## Electron version 0.37.6 (process.versions.chrome = 49.0.2623.75)
image

## More info:
* **Reproduces** in Electron version 0.37.6
* fyi @bbondy Reproduces in Brave 0.12.0 (Electron 1.3.10, libchromiumcontent: 52.0.2743.116)
* Reproduces on Surface Book, in external monitor at 100% zoom
* **Does not** reproduce on Lenovo w530 with the same Windows version
* this impacts VSCode considerably (why I started looking into this)

## Hints:
* The only hints I could find are when clicking on the input events and most of them in Electron appear to be waiting a long time on the main process. For Chromium 52 or Chrome 53, the total time of these events is never longer than 20-30ms, typically even shorter:

image

fyi @zcbenz @kevinsawicki @bpasero @dbaeumer

### PR DESCRIPTION
**Pull request by paulcbetts**
Fixes #7208 

## TODO:

- [x] Verify with a touch screen
- [x] Verify that NVDA still works

### CODE REVIEW

File: atom/browser/native_window_views.h [hunk 1]
```diff
@@ -18,6 +18,8 @@
 #include "atom/browser/ui/win/message_handler_delegate.h"
 #include "atom/browser/ui/win/taskbar_host.h"
 #include "base/win/scoped_gdi_object.h"
+#include "ui/base/win/accessibility_misc_utils.h"
```
Comment (kevinsawicki) [inline atom/browser/native_window_views.h]: Just curious, what is this include used for?

Comment (paulcbetts) [inline atom/browser/native_window_views.h]: @kevinsawicki Without it, the next header barfs a thousand lines of COM nonsense

File: atom/browser/native_window_views_win.cc [hunk 2]
```diff
@@ -92,15 +92,25 @@ bool NativeWindowViews::PreHandleMSG(
...
-      if (obj_id == OBJID_CLIENT) {
-        const auto axState = content::BrowserAccessibilityState::GetInstance();
-        if (axState && !axState->IsAccessibleBrowser()) {
-          axState->OnScreenReaderDetected();
-          enabled_a11y_support_ = true;
-          Browser::Get()->OnAccessibilitySupportChanged();
-        }
+      if (checked_for_a11y_support_) return false;
+
+      checked_for_a11y_support_ = true;
```
Comment (kevinsawicki) [inline atom/browser/native_window_views_win.cc]: On this branch (compared with master), I'm not seeing the `'accessibility-support-changed'` event fire when you launch NVDA after launching your app:

To reproduce:

- Launch Electron
- Open the dev tools and run the following
```js
require('electron').remote.app.on('accessibility-support-changed', function(e, axs) {console.log('changed', axs)})
```
- Launch NVDA
- 'changed true` is not logged on this branch (it is logged on master)

Should the `checked_for_a11y_support_ = true;` be done after the `if (obj_id != OBJID_CLIENT) {` check? If I move it below that check, then I do see it fire.

File: atom/browser/native_window_views.h [hunk 3]
```diff
@@ -228,6 +230,9 @@ class NativeWindowViews : public NativeWindow,
   // In charge of running taskbar related APIs.
   TaskbarHost taskbar_host_;
 
+  // Memoized version of a11y check
+  bool checked_for_a11y_support_;
```
Comment (kevinsawicki) [inline atom/browser/native_window_views.h]: This should be initialized to `false` in the constructor like `enabled_a11y_support_` is in [`atom/browser/native_window_views.cc`](https://github.com/electron/electron/blob/988e2334f59d8aaccdec394a9d9fa0c39920068d/atom/browser/native_window_views.cc#L141)

File: atom/browser/native_window_views_win.cc [hunk 4]
```diff
@@ -92,15 +92,25 @@ bool NativeWindowViews::PreHandleMSG(
...
+      UINT screenReader = 0;
+      SystemParametersInfo(SPI_GETSCREENREADER, 0, &screenReader, 0);
+      if (!screenReader || !UiaClientsAreListening()) {
+        return false;
+      }
+
+      const auto axState = content::BrowserAccessibilityState::GetInstance();
+      if (axState && !axState->IsAccessibleBrowser()) {
+        axState->OnScreenReaderDetected();
+        enabled_a11y_support_ = true;
```
Comment (kevinsawicki) [inline atom/browser/native_window_views_win.cc]: I think this variable can be removed now, it looks like it is no longer read anywhere.

File: atom/browser/native_window_views_win.cc (hunk 2 above)
Comment (paulcbetts) [inline atom/browser/native_window_views_win.cc]: We can do that, I was trying to make this as efficient as possible but I didn't know we had an event for this, :+1:

File: atom/browser/native_window_views_win.cc (hunk 4 above)
Comment (paulcbetts) [inline atom/browser/native_window_views_win.cc]: :+1:

**Comment by kevinsawicki** [comment]
Thanks so much for fixing this 💻 🔍

**Comment by miniak** [comment]
@paulcbetts, why didn't you do it like this?

```
static bool IsScreenReaderActive()
{
  UINT screenReader = 0;
  SystemParametersInfo(SPI_GETSCREENREADER, 0, &screenReader, 0);

  return screenReader && UiaClientsAreListening();
}

      if (obj_id == OBJID_CLIENT && IsScreenReaderActive()) {
```

That would be much cleaner.

**Comment by miniak** [comment]
@paulcbetts did you try running Electron first and then running the screen reader? You only check if it's running once.

**Comment by kevinsawicki** [comment]
>  did you try running Electron first and then running the screen reader?

I did try this and it worked as expected but I wasn't on a touch screen. I launched Electron, then opened NVDA, and saw the accessibility state changed even fire.

**Comment by miniak** [comment]
@kevinsawicki, I am experimenting with this. Looks like the whole WM_GETOBJECT handling code may not be even necessary. JAWS, NVDA and Window-Eyes trigger the accessible mode without that. Narrator and ZoomText only work if they were running before Electron was launched.

**Comment by miniak** [comment]
Also the `accessibility-support-changed` event should be fired when `BrowserAccessibilityState::IsAccessibleBrowser()` value changes, `BrowserAccessibilityState::OnScreenReaderDetected()` does not have to be called by Electron code (https://github.com/adobe/chromium/blob/cfe5bf0b51b1f6b9fe239c2a3c2f2364da9967d7/content/browser/renderer_host/render_widget_host_view_win.cc#L2307)

**Comment by miniak** [comment]
Unfortunately there is not notification for `BrowserAccessibilityState::IsAccessibleBrowser()` changes

**Comment by miniak** [comment]
Also, on my Windows 10 in Parallels on Mac, `UiaClientsAreListening` returns true even if I don't have any screen reader running.

**Comment by alexandrudima** [comment]
@paulcbetts Thank you very much for the fix! :heart:

**Pull request merged**